airbyte-integrations/connectors/source-marketo/source_marketo/source.py

```diff
@@ -0,0 +1,530 @@
+            "type": ["null", "object"],
+            "additionalProperties": False,
+            "properties": properties,
+        }
+
+        return schema
+
+
+class ActivityTypes(MarketoStream):
+    """
+    Return list of all activity types.
+    API Docs: http://developers.marketo.com/rest-api/lead-database/activities/#describe
+    """
+
+    def path(self, stream_slice: Mapping[str, Any] = None, **kwargs) -> str:
+        return "/rest/v1/activities/types.json"
+
+
+class Programs(IncrementalMarketoStream):
+    """
+    Return list of all programs.
+    API Docs: http://developers.marketo.com/rest-api/assets/programs/#by_date_range
+    """
+
+    cursor_field = "updatedAt"
+    page_size = 200
+    offset = 0
+
+    def path(self, **kwargs) -> str:
+        return f"/rest/asset/v1/{self.name}.json"
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        data = response.json().get(self.data_field)
+
+        if data:
+            self.offset += self.page_size + 1
+            return {"offset": self.offset}
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, Any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        """
+        Programs are queryable via their updatedAt time but require and
+        end date as well. As there is no max time range for the query,
+        query from the bookmark value until current.
+        """
+
+        params = super().request_params(stream_state, stream_slice, next_page_token)
+        params.update(
+            {
+                "maxReturn": self.page_size,
+                "earliestUpdatedAt": stream_slice["startAt"],
+                "latestUpdatedAt": stream_slice["endAt"],
+            }
+        )
+
+        return params
+
+    def parse_response(self, response: requests.Response, stream_state: Mapping[str, Any], **kwargs) -> Iterable[Mapping]:
+        for record in super().parse_response(response, stream_state, **kwargs):
+            record["updatedAt"] = record["updatedAt"][:-5]
+            yield record
+
+
+class Campaigns(IncrementalMarketoStream):
+    """
+    Return list of all campaigns.
+    API Docs: http://developers.marketo.com/rest-api/endpoint-reference/lead-database-endpoint-reference/#!/Campaigns/getCampaignsUsingGET
+    """
+
+    pass
+
+
+class Lists(IncrementalMarketoStream):
+    """
+    Return list of all lists.
+    API Docs: http://developers.marketo.com/rest-api/endpoint-reference/lead-database-endpoint-reference/#!/Static_Lists/getListsUsingGET
+    """
+
+    pass
+
+
+class MarketoAuthenticator(Oauth2Authenticator):
+    def __init__(self, config):
+        super().__init__(
+            token_refresh_endpoint=f"{config['base_url']}/identity/oauth/token",
+            client_id=config["client_id"],
+            client_secret=config["client_secret"],
+            refresh_token=None,
+        )
```

Comment:
Please override only the necessary parts, It could be much less code

Reply:
Already overrided only the necessary parts.